File: sdks/javascript/examples/iot-sensor-pipeline.ts
/**
 * IoT Sensor Pipeline Example (TypeScript)
 *
 * Demonstrates:
 * - TumblingWindow for time-based aggregation of sensor events
 * - Watermark-based late data handling
 * - BackpressureController for burst sensor data
 * - MultiLevelBackpressure with priority queues
 * - RateBasedBackpressure for rate limiting
 * - Alert generation when thresholds are exceeded
 */

import {
  Timestamp,
  Duration,
  StreamEvent,
  Watermark,
  WindowType,
  PaneInfo,
  BackpressureStrategy,
  LateDataPolicy,
  createStreamEvent,
  createWindowSpec,
  createWindowInfo,
  createStreamConfig,
  createBackpressureConfig,
} from '../src/streaming/types';
import { TumblingWindow, WindowAssigner, WindowTrigger } from '../src/streaming/window';
import { BackpressureController, MultiLevelBackpressure, RateBasedBackpressure } from '../src/streaming/backpressure';

// ========================================
// Types
// ========================================

interface SensorReading {
  sensorId: string;
  sensorType: string;
  value: number;
  unit: string;
}

interface AggregatedWindow {
  sensorType: string;
  avg: number;
  min: number;
  max: number;
  count: number;
  windowStartMs: number;
  windowEndMs: number;
}

// ========================================
// Constants
// ========================================

const THRESHOLDS: Record<string, { min: number; max: number; unit: string }> = {
  temperature: { min: -10, max: 50, unit: 'C' },
  humidity: { min: 0, max: 100, unit: '%' },
  pressure: { min: 950, max: 1050, unit: 'hPa' },
};

const SENSOR_IDS: Record<string, string[]> = {
  temperature: ['temp-001', 'temp-002'],
  humidity: ['hum-001', 'hum-002'],
  pressure: ['press-001'],
};

const RANGES: Record<string, [number, number]> = {
  temperature: [15, 35],
  humidity: [30, 80],
  pressure: [990, 1030],
};

// ========================================
// Utilities
// ========================================

const AUDIT_LOG: string[] = [];

function log(msg: string): void {
  const ts = new Date().toISOString().substring(11, 23);
  const entry = `  [${ts}] ${msg}`;
  AUDIT_LOG.push(entry);
  console.log(entry);
}

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function randomRange(min: number, max: number): number {
  return Math.round((Math.random() * (max - min) + min) * 10) / 10;
}

function randomChoice<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

function randomChoiceWeighted<T>(items: T[], weights: number[]): T {
  const total = weights.reduce((a, b) => a + b, 0);
  let r = Math.random() * total;
  for (let i = 0; i < items.length; i++) {
    r -= weights[i];
    if (r <= 0) return items[i];
  }
  return items[items.length - 1];
}

// ========================================
// Sensor helpers
// ========================================

function generateReading(sensorType: string): SensorReading {
  const [lo, hi] = RANGES[sensorType];
  return {
    sensorId: randomChoice(SENSOR_IDS[sensorType]),
    sensorType,
    value: randomRange(lo, hi),
    unit: THRESHOLDS[sensorType].unit,
  };
}

function checkThreshold(reading: SensorReading): string | null {
  const t = THRESHOLDS[reading.sensorType];
  if (reading.value > t.max) {
    return `HIGH: ${reading.sensorType}=${reading.value}${t.unit} exceeds ${t.max}${t.unit}`;
  }
  if (reading.value < t.min) {
    return `LOW: ${reading.sensorType}=${reading.value}${t.unit} below ${t.min}${t.unit}`;
  }
  return null;
}

function aggregateEvents(
  events: StreamEvent<SensorReading>[],
  info: { start: Timestamp; end: Timestamp; maxTimestamp: Timestamp; pane: PaneInfo; windowId?: string }
): AggregatedWindow {
  if (events.length === 0) {
    return { sensorType: '', avg: 0, min: 0, max: 0, count: 0, windowStartMs: 0, windowEndMs: 0 };
  }

  const values = events.map(e => e.value.value);
  const sensorType = events[0].value.sensorType;

  return {
    sensorType,
    avg: Math.round((values.reduce((a, b) => a + b, 0) / values.length) * 10) / 10,
    min: Math.min(...values),
    max: Math.max(...values),
    count: values.length,
    windowStartMs: info.start.milliseconds,
    windowEndMs: info.end.milliseconds,
  };
}

// ========================================
// Main
// ========================================

async function runSensorPipeline(): Promise<void> {
  console.log('='.repeat(70));
  console.log('  AETHER IOT SENSOR PIPELINE (TypeScript)');
  console.log('='.repeat(70));
  console.log();

  const windowSize = Duration.fromSeconds(60);
  const allAggregations: AggregatedWindow[] = [];
  const allAlerts: string[] = [];

  console.log('--- Step 1: Set up tumbling windows per sensor type ---');
  const tempWindow = new TumblingWindow<string, SensorReading, AggregatedWindow>(
    windowSize, aggregateEvents
  );
  const humWindow = new TumblingWindow<string, SensorReading, AggregatedWindow>(
    windowSize, aggregateEvents
  );
  const pressWindow = new TumblingWindow<string, SensorReading, AggregatedWindow>(
    windowSize, aggregateEvents
  );
  log(`Created tumbling windows (size=${windowSize.toSeconds()}s)`);
  console.log();

  console.log('--- Step 2: Set up backpressure controller ---');
  const bpConfig = createBackpressureConfig({
    strategy: BackpressureStrategy.Buffer,
    bufferSize: 50,
    highWatermark: 0.8,
    lowWatermark: 0.3,
  });
  const backpressure = new BackpressureController<SensorReading>(bpConfig);
  let overflowCount = 0;
  let resumeCount = 0;

  backpressure.setOverflowCallback(() => {
    overflowCount++;
    log(`BACKPRESSURE: High watermark reached (overflow #${overflowCount})`);
  });
  backpressure.setResumeCallback(() => {
    resumeCount++;
    log(`BACKPRESSURE: Buffer recovered (resume #${resumeCount})`);
  });

  log(`Backpressure configured: strategy=${bpConfig.strategy}, buffer=${bpConfig.bufferSize}`);
  console.log();

  console.log('--- Step 3: Set up watermark tracking ---');
  const watermarks: Map<string, Timestamp> = new Map();
  const lateEvents: StreamEvent<SensorReading>[] = [];
  log('Watermark tracking initialized');
  console.log();

  console.log('--- Step 4: Generate and process sensor events ---');
  console.log();

  const baseTimeMs = Math.floor(Date.now() / 60000) * 60000;
  const windowMap: Record<string, TumblingWindow<string, SensorReading, AggregatedWindow>> = {
    temperature: tempWindow,
    humidity: humWindow,
    pressure: pressWindow,
  };

  let eventsAccepted = 0;
  let eventsDropped = 0;
  let eventsLate = 0;

  for (let secondOffset = 0; secondOffset < 125; secondOffset += 2) {
    const eventTimeMs = baseTimeMs + secondOffset * 1000;
    const eventTs = new Timestamp(eventTimeMs);
    const numReadings = randomChoiceWeighted([1, 2, 3], [5, 3, 1]);

    for (let r = 0; r < numReadings; r++) {
      const sensorType = randomChoice(['temperature', 'humidity', 'pressure']);

      const isLate = Math.random() < 0.05;
      if (isLate && secondOffset > 30) {
        const reading = generateReading(sensorType);
        const lateTs = new Timestamp(eventTimeMs - Math.floor(Math.random() * 30000 + 90000));
        const lateEvent = createStreamEvent<SensorReading>(
          reading.sensorId,
          reading,
          lateTs,
          { eventType: sensorType },
        );

        const currentWm = watermarks.get(sensorType) ?? new Timestamp(0);
        if (lateEvent.timestamp.isBefore(currentWm)) {
          eventsLate++;
          lateEvents.push(lateEvent);
          const lateTime = lateEvent.timestamp.toDate().toISOString().substring(11, 19);
          const wmTime = currentWm.toDate().toISOString().substring(11, 19);
          log(`LATE DATA: ${sensorType} from ${reading.sensorId} at ${lateTime} (watermark: ${wmTime})`);
          continue;
        }
      }

      const reading = generateReading(sensorType);
      const event = createStreamEvent<SensorReading>(
        reading.sensorId,
        reading,
        eventTs,
        { eventType: sensorType },
      );

      const alert = checkThreshold(reading);
      if (alert) {
        allAlerts.push(alert);
        log(`ALERT: ${alert}`);
      }

      if (!backpressure.tryPush(event)) {
        eventsDropped++;
        log(`DROPPED: Event from ${reading.sensorId} (buffer full)`);
        continue;
      }

      eventsAccepted++;

      const bufferedEvent = backpressure.pop();
      if (bufferedEvent) {
        const window = windowMap[(bufferedEvent.value as unknown as SensorReading).sensorType];
        const results = window.process(bufferedEvent, bufferedEvent.eventType ?? 'default');
        allAggregations.push(...results);
      }
    }

    const watermarkTs = new Timestamp(eventTimeMs + 5000);
    for (const stype of ['temperature', 'humidity', 'pressure']) {
      watermarks.set(stype, watermarkTs);
      const window = windowMap[stype];
      const results = window.advanceWatermark(watermarkTs);
      allAggregations.push(...results);
    }

    if (secondOffset % 20 === 0) {
      const stats = backpressure.getStats();
      log(`PROGRESS: t=${secondOffset}s | accepted=${eventsAccepted} ` +
        `dropped=${eventsDropped} late=${eventsLate} ` +
        `buffer=${stats.currentBufferSize}/${bpConfig.bufferSize} ` +
        `windows_fired=${allAggregations.length}`);
    }
  }

  console.log();

  console.log('--- Step 5: Fire remaining windows ---');
  const finalWatermark = new Timestamp(baseTimeMs + 180000);
  for (const [stype, window] of Object.entries(windowMap)) {
    const results = window.advanceWatermark(finalWatermark);
    allAggregations.push(...results);
  }
  log(`Fired remaining windows, total aggregations: ${allAggregations.length}`);
  console.log();

  console.log('--- Step 6: Multi-level backpressure demo ---');
  const mlbp = new MultiLevelBackpressure<SensorReading>(10);
  for (let i = 0; i < 15; i++) {
    const priority = i < 8 ? MultiLevelBackpressure.LOW : MultiLevelBackpressure.HIGH;
    const event = createStreamEvent(`sensor-${i}`, { sensorId: `${i}`, sensorType: 'temp', value: i, unit: 'C' });
    const accepted = mlbp.push(event, priority);
    if (!accepted) {
      log(`MLBP: Event ${i} dropped (priority=${priority})`);
    }
  }
  const mlbpStats = mlbp.getStats();
  log(`Multi-level BP: total=${mlbpStats.totalEvents}, dropped=${mlbpStats.droppedEvents}, buffered=${mlbpStats.bufferedEvents}`);
  while (!mlbp.isEmpty()) {
    mlbp.pop();
  }
  log('Multi-level BP: drained');
  console.log();

  console.log('--- Step 7: Rate-based backpressure demo ---');
  const rateBP = new RateBasedBackpressure(5, 1.0, 0.2);
  let allowed = 0;
  let rejected = 0;
  for (let i = 0; i < 20; i++) {
    if (await rateBP.tryAcquire()) {
      allowed++;
    } else {
      rejected++;
    }
  }
  log(`Rate-based BP: allowed=${allowed}, rejected=${rejected}, active=${rateBP.isBackpressureActive}`);
  console.log();

  console.log('='.repeat(70));
  console.log('  PIPELINE SUMMARY');
  console.log('='.repeat(70));
  log(`Total events accepted: ${eventsAccepted}`);
  log(`Total events dropped:  ${eventsDropped}`);
  log(`Total late events:     ${eventsLate}`);
  log(`Total alerts:          ${allAlerts.length}`);
  log(`Total window results:  ${allAggregations.length}`);
  log(`Backpressure overflows: ${overflowCount}`);
  log(`Backpressure resumes:   ${resumeCount}`);
  console.log();

  if (allAggregations.length > 0) {
    console.log('  Window Aggregations:');
    for (const agg of allAggregations.slice(0, 12)) {
      const durationS = (agg.windowEndMs - agg.windowStartMs) / 1000;
      log(`  [${agg.sensorType.padEnd(11)}] avg=${String(agg.avg).padStart(6)} ` +
        `min=${String(agg.min).padStart(6)} max=${String(agg.max).padStart(6)} ` +
        `count=${String(agg.count).padStart(3)} window=${durationS}s`);
    }
    if (allAggregations.length > 12) {
      log(`  ... and ${allAggregations.length - 12} more`);
    }
  }
  console.log();

  if (allAlerts.length > 0) {
    console.log('  Alerts Generated:');
    for (const alert of allAlerts.slice(0, 8)) {
      log(`  ! ${alert}`);
    }
    if (allAlerts.length > 8) {
      log(`  ... and ${allAlerts.length - 8} more`);
    }
  } else {
    log('No threshold alerts generated (all readings within range)');
  }
  console.log();

  log(`Total audit entries: ${AUDIT_LOG.length}`);
}

(async () => {
  await runSensorPipeline();
})();

File: sdks/python/examples/iot_sensor_pipeline.py
"""
IoT Sensor Pipeline Example

Demonstrates:
- Tumbling windows for time-based aggregation of sensor events
- Watermark-based late data handling
- Backpressure for burst sensor data
- Alert generation when thresholds are exceeded
- StreamEvent processing with StreamActor

Simulates three sensor types (temperature, humidity, pressure) sending
events at realistic intervals through a pipeline that aggregates
per-minute windows and generates alerts.
"""

import asyncio
import random
import time
from dataclasses import dataclass
from typing import Dict, List, Optional

from aether_sdk.streaming.backpressure import (
    BackpressureController,
    MultiLevelBackpressure,
    RateBasedBackpressure,
)
from aether_sdk.streaming.types import (
    BackpressureConfig,
    BackpressureStrategy,
    Duration,
    StreamEvent,
    Timestamp,
    WindowInfo,
)
from aether_sdk.streaming.window import TumblingWindow

AUDIT_LOG = []


def log(msg: str):
    ts = time.strftime("%H:%M:%S")
    AUDIT_LOG.append(f"[{ts}] {msg}")
    print(f"  [{ts}] {msg}")


# ========================================
# Sensor Event Types
# ========================================


@dataclass
class SensorReading:
    sensor_id: str
    sensor_type: str
    value: float
    unit: str


THRESHOLDS = {
    "temperature": {"min": -10.0, "max": 50.0, "unit": "C"},
    "humidity": {"min": 0.0, "max": 100.0, "unit": "%"},
    "pressure": {"min": 950.0, "max": 1050.0, "unit": "hPa"},
}

SENSOR_IDS = {
    "temperature": ["temp-001", "temp-002"],
    "humidity": ["hum-001", "hum-002"],
    "pressure": ["press-001"],
}


def generate_reading(sensor_type: str) -> SensorReading:
    ranges = {
        "temperature": (15.0, 35.0),
        "humidity": (30.0, 80.0),
        "pressure": (990.0, 1030.0),
    }
    low, high = ranges[sensor_type]
    value = round(random.uniform(low, high), 1)
    sensor_id = random.choice(SENSOR_IDS[sensor_type])
    return SensorReading(
        sensor_id=sensor_id,
        sensor_type=sensor_type,
        value=value,
        unit=THRESHOLDS[sensor_type]["unit"],
    )


def check_threshold(reading: SensorReading) -> Optional[str]:
    threshold = THRESHOLDS[reading.sensor_type]
    if reading.value > threshold["max"]:
        return f"HIGH: {reading.sensor_type}={reading.value}{threshold['unit']} exceeds {threshold['max']}{threshold['unit']}"
    if reading.value < threshold["min"]:
        return f"LOW: {reading.sensor_type}={reading.value}{threshold['unit']} below {threshold['min']}{threshold['unit']}"
    return None


# ========================================
# Window Aggregation
# ========================================


@dataclass
class AggregatedWindow:
    sensor_type: str
    avg: float
    min_val: float
    max_val: float
    count: int
    window_start: int
    window_end: int


def aggregate_events(
    events: List[StreamEvent[SensorReading]], window_info: WindowInfo
) -> AggregatedWindow:
    if not events:
        return AggregatedWindow(
            sensor_type="",
            avg=0,
            min_val=0,
            max_val=0,
            count=0,
            window_start=0,
            window_end=0,
        )

    values = [e.value.value for e in events]
    sensor_type = events[0].value.sensor_type

    return AggregatedWindow(
        sensor_type=sensor_type,
        avg=round(sum(values) / len(values), 1),
        min_val=min(values),
        max_val=max(values),
        count=len(values),
        window_start=window_info.start.milliseconds,
        window_end=window_info.end.milliseconds,
    )


# ========================================
# Pipeline
# ========================================


async def run_sensor_pipeline():
    print("=" * 70)
    print("  AETHER IOT SENSOR PIPELINE")
    print("=" * 70)
    print()

    window_size = Duration.from_seconds(60)
    all_aggregations: List[AggregatedWindow] = []
    all_alerts: List[str] = []

    print("--- Step 1: Set up tumbling windows per sensor type ---")
    temp_window = TumblingWindow(window_size, aggregate_events)
    hum_window = TumblingWindow(window_size, aggregate_events)
    press_window = TumblingWindow(window_size, aggregate_events)
    log(f"Created tumbling windows (size={window_size.to_seconds()}s)")
    print()

    print("--- Step 2: Set up backpressure controller ---")
    bp_config = BackpressureConfig(
        strategy=BackpressureStrategy.BUFFER,
        buffer_size=50,
        high_watermark=0.8,
        low_watermark=0.3,
    )
    backpressure = BackpressureController(bp_config)
    overflow_count = [0]

    def on_overflow():
        overflow_count[0] += 1
        log(f"BACKPRESSURE: High watermark reached (overflow #{overflow_count[0]})")

    backpressure.set_overflow_callback(on_overflow)

    resume_count = [0]

    def on_resume():
        resume_count[0] += 1
        log(f"BACKPRESSURE: Buffer recovered (resume #{resume_count[0]})")

    backpressure.set_resume_callback(on_resume)
    log(
        f"Backpressure configured: strategy={bp_config.strategy.value}, buffer={bp_config.buffer_size}"
    )
    print()

    print("--- Step 3: Set up watermark tracking ---")
    watermarks: Dict[str, Timestamp] = {}
    late_events: List[StreamEvent] = []

    log("Watermark tracking initialized")
    print()

    print("--- Step 4: Generate and process sensor events ---")
    print()

    base_time_ms = int(time.time() * 1000)
    base_time_ms = (base_time_ms // 60000) * 60000

    window_map = {
        "temperature": temp_window,
        "humidity": hum_window,
        "pressure": press_window,
    }

    events_accepted = 0
    events_dropped = 0
    events_late = 0

    for second_offset in range(0, 125, 2):
        event_time_ms = base_time_ms + (second_offset * 1000)
        event_ts = Timestamp(event_time_ms)

        num_readings = random.choices([1, 2, 3], weights=[5, 3, 1])[0]

        for _ in range(num_readings):
            sensor_type = random.choice(["temperature", "humidity", "pressure"])

            is_late = random.random() < 0.05
            if is_late and second_offset > 30:
                reading = generate_reading(sensor_type)
                late_ts = Timestamp(event_time_ms - random.randint(90000, 120000))
                late_event = StreamEvent.create(
                    key=reading.sensor_id,
                    value=reading,
                    timestamp=late_ts,
                    event_type=sensor_type,
                )

                current_wm = watermarks.get(sensor_type, Timestamp(0))
                if late_event.timestamp < current_wm:
                    events_late += 1
                    late_events.append(late_event)
                    log(
                        f"LATE DATA: {sensor_type} from {reading.sensor_id} "
                        f"at {late_ts.to_datetime().strftime('%H:%M:%S')} "
                        f"(watermark: {current_wm.to_datetime().strftime('%H:%M:%S')})"
                    )
                    continue

            reading = generate_reading(sensor_type)
            event = StreamEvent.create(
                key=reading.sensor_id,
                value=reading,
                timestamp=event_ts,
                event_type=sensor_type,
            )

            alert = check_threshold(reading)
            if alert:
                all_alerts.append(alert)
                log(f"ALERT: {alert}")

            if not backpressure.try_push(event):
                events_dropped += 1
                log(f"DROPPED: Event from {reading.sensor_id} (buffer full)")
                continue

            events_accepted += 1

            buffered_event = backpressure.pop()
            if buffered_event is not None:
                window = window_map[buffered_event.value.sensor_type]
                results = window.process(buffered_event, buffered_event.event_type)
                for r in results:
                    all_aggregations.append(r)

        watermark_ts = Timestamp(event_time_ms + 5000)
        for stype in ["temperature", "humidity", "pressure"]:
            watermarks[stype] = watermark_ts
            window = window_map[stype]
            results = window.advance_watermark(watermark_ts)
            for r in results:
                all_aggregations.append(r)

        if second_offset % 20 == 0:
            stats = backpressure.stats
            log(
                f"PROGRESS: t={second_offset}s | accepted={events_accepted} "
                f"dropped={events_dropped} late={events_late} "
                f"buffer={stats.current_buffer_size}/{bp_config.buffer_size} "
                f"windows_fired={len(all_aggregations)}"
            )

    print()

    print("--- Step 5: Fire remaining windows ---")
    final_watermark = Timestamp(base_time_ms + 180000)
    for stype, window in window_map.items():
        results = window.advance_watermark(final_watermark)
        for r in results:
            all_aggregations.append(r)
    log(f"Fired remaining windows, total aggregations: {len(all_aggregations)}")
    print()

    print("--- Step 6: Multi-level backpressure demo ---")
    mlbp = MultiLevelBackpressure(buffer_size=10)
    for i in range(15):
        priority = (
            MultiLevelBackpressure.Priority.LOW
            if i < 8
            else MultiLevelBackpressure.Priority.HIGH
        )
        event = StreamEvent.create(
            key=f"sensor-{i}", value=i, timestamp=Timestamp.now()
        )
        accepted = mlbp.push(event, priority)
        if not accepted:
            log(f"MLBP: Event {i} dropped (priority={priority})")

    mlbp_stats = mlbp.size()
    log(f"Multi-level BP: buffered={mlbp_stats}, " f"dropped={15 - mlbp_stats}")

    while not mlbp.is_empty():
        mlbp.pop()
    log("Multi-level BP: drained")
    print()

    print("--- Step 7: Rate-based backpressure demo ---")
    rate_bp = RateBasedBackpressure(max_rate=5, window_size=1.0, cooldown=0.2)
    allowed = 0
    rejected = 0
    for _ in range(20):
        if await rate_bp.try_acquire():
            allowed += 1
        else:
            rejected += 1
    log(
        f"Rate-based BP: allowed={allowed}, rejected={rejected}, "
        f"active={rate_bp.is_backpressure_active}"
    )
    print()

    print("=" * 70)
    print("  PIPELINE SUMMARY")
    print("=" * 70)
    log(f"Total events accepted: {events_accepted}")
    log(f"Total events dropped:  {events_dropped}")
    log(f"Total late events:     {events_late}")
    log(f"Total alerts:          {len(all_alerts)}")
    log(f"Total window results:  {len(all_aggregations)}")
    log(f"Backpressure overflows: {overflow_count[0]}")
    log(f"Backpressure resumes:   {resume_count[0]}")
    print()

    if all_aggregations:
        print("  Window Aggregations:")
        for agg in all_aggregations[:12]:
            duration_s = (agg.window_end - agg.window_start) / 1000
            log(
                f"  [{agg.sensor_type:11s}] avg={agg.avg:6.1f} "
                f"min={agg.min_val:6.1f} max={agg.max_val:6.1f} "
                f"count={agg.count:3d} window={duration_s:.0f}s"
            )
        if len(all_aggregations) > 12:
            log(f"  ... and {len(all_aggregations) - 12} more")
    print()

    if all_alerts:
        print("  Alerts Generated:")
        for alert in all_alerts[:8]:
            log(f"  ! {alert}")
        if len(all_alerts) > 8:
            log(f"  ... and {len(all_alerts) - 8} more")
    else:
        log("No threshold alerts generated (all readings within range)")
    print()

    print(f"Total audit entries: {len(AUDIT_LOG)}")


if __name__ == "__main__":
    asyncio.run(run_sensor_pipeline())
